api/agent/runner_client.go

```diff
@@ -199,7 +199,11 @@ func receiveFromRunner(protocolClient pb.RunnerProtocol_EngageClient, c pool.Run
 			if body.Finished.Success {
 				logrus.Infof("Call finished successfully: %v", body.Finished.Details)
 			} else {
-				logrus.Infof("Call finish unsuccessfully:: %v", body.Finished.Details)
+				logrus.Infof("Call finished unsuccessfully: %v", body.Finished.Details)
+			}
+			// There should be an EOF following the last packet
+			if _, err := protocolClient.Recv(); err != io.EOF {
+				logrus.Errorf("Did not receive expected EOF from runner stream: %v", err)
```

Comment:
Capture error with `logrus.WithError(err).Error("Did not receive expected EOF from runner stream")`

Reply:
We might also want to explicitly post the error via the channel, since receiving other messages potentially indicates a bug or issue with the state of the runner protocol messages:

`done <- error`

Reply:
Good idea; doing now.

Reply:
In fact, I think that `pure_runner.go` should be modified to do a `SendAndClose` on sending a RunnerMsg_Finished message.